### ISSUE TITLE
Ajax subscribe to a specific list

### ISSUE THREAD

**Issue by bramley** [phpList/phplist3#701]
The ajax subscribe process currently subscribes the user to all lists associated with the subscribe page, and sets the email format to html

                foreach (explode(',', $GLOBALS['pagedata']['lists']) as $listid) {
                    $_POST['list'][$listid] = 'signup';
                }
                $_POST['htmlemail'] = 1; //# @@ should actually be taken from the subscribe page data

A problem raised in the forum  https://discuss.phplist.org/t/is-it-possible-to-subscribe-to-specific-list-from-subscription-page-through-ajax/6628 asks whether the ajax request could specify a particular list instead of all lists being used. It looks a simple change to only assign all lists when one has not been included in the request (`$_POST['list']` not being set)

Similarly, set the email format to html, or that specified in the subscribe page itself, only when it is not included in the request.

@michield How do you feel about this? Would it adversely affect people who use ajax with the hosted service? As long as the request does not include list or htmlemail then they should have the same result.

**Comment by michield** [comment]
If I remember correctly the choice of lists is decided by the subscribe page. So, if you create a subscribe page with just the lists you want them to sign up to, it will sign them up to all lists of that page. That way you can control which lists are used.

**Comment by michield** [comment]
As for html format, doesn't everyone just want html now? Not sure why you would make it a choice.

**Comment by bramley** [comment]
I think that the concern is siging up to several lists offered by a subscribe page instead of just one list. The forum topic that I linked to tries to explain the issue.
Say you have lists for Java, PHP and C++ offered on one subscribe page. The phplist generated subscribe page will show all of the three lists. That's fine, the person can select which list they are interested in. Later, on the preferences page they can signup to another list.

What happens currently with ajax is that someone is subscribed to all of the lists, whereas he wants to have individual signup forms for each list, e.g. "sign up to our PHP weekly newsletter" ajax form, "signup to our weekly Java newsletter" form, etc. If someone can do that then they can discover the other lists through the preferences page.

If instead there are separate subscribe pages for each language then the preferences page won't show the other languages, so the subscriber won't be aware that there are other lists available.

**Comment by michield** [comment]
Ah, yes, I see what you mean. I guess we can add an option to specify the list in the ajax call. If we make it a new var, it should work with backwards compatibility. Are you happy to sort that out?

**Comment by bramley** [comment]
I was thinking of simply using the same $_POST variable so that it remains the same as the subscribe page.
The ajax form would include something similar to this to subscribe to list id 2

    <input type="hidden" name="list[2]" value="signup"/>

and amend index.php

                if (!isset($_POST['list'])) {
                    foreach (explode(',', $GLOBALS['pagedata']['lists']) as $listid) {
                        $_POST['list'][$listid] = 'signup';
                    }
                }

If you want to have a new $_POST field

    <input type="hidden" name="listid" value="2"/>

and amend index.php

                if (isset($_POST['listid'])) {
                    $listid = $_POST['listid'];
                    $_POST['list'][$listid] = 'signup';
                } else {
                    foreach (explode(',', $GLOBALS['pagedata']['lists']) as $listid) {
                        $_POST['list'][$listid] = 'signup';
                    }
                }

**Comment by michield** [comment]
Yes, option 1 looks fine. You'd have to hard code the list IDs, but that can be explained in the docs.

If you do the second one, please make sure to sanitise the post data

                $listid = sprintf('%d',$_POST['listid']);
